src/pkg/cli/client/byoc/aws/byoc.go

```diff
@@ -322,81 +322,99 @@ func (b *ByocAws) findZone(ctx context.Context, domain, roleARN string) (string,
 		}
 		term.Debugf("Zone %q not found, delegation set will be created", projectDomain)
-		// Case 1: The zone doesn't exist: we'll create a delegation set and let CD/Pulumi create the hosted zone
+
+		// Case 1: The zone doesn't exist: we'll create/get a delegation set and let CD/Pulumi create the hosted zone
+
+		// Avoid creating a new delegation set if one already exists
+		delegationSet, err = aws.GetDelegationSet(ctx, r53Client)
+		// Create a new delegation set if it doesn't exist
+		if errors.Is(err, aws.ErrNoDelegationSetFound) {
+			// Create a new delegation set. There's a race condition here, where two deployments could create two different delegation sets,
+			// but this is acceptable because the next time the zone is deployed, we'll get the existing delegation set from the zone.
+			delegationSet, err = aws.CreateDelegationSet(ctx, nil, r53Client)
+		}
+		if err != nil {
+			return nil, "", err
+		}
 	} else {
 		// Case 2: Get the NS records for the existing subdomain zone
 		nsServers, err = aws.ListResourceRecords(ctx, *zone.Id, projectDomain, r53types.RRTypeNs, r53Client)
 		if err != nil {
-			return nil, AnnotateAwsError(err) // TODO: we should not fail deployment if this fails
+			return nil, "", err // TODO: we should not fail deployment if this fails
 		}
 		term.Debugf("Zone %q found, NS records: %v", projectDomain, nsServers)
-	}
 
-	var resp client.PrepareDomainDelegationResponse
-	if zone == nil || zone.Config.Comment == nil || *zone.Config.Comment != aws.CreateHostedZoneComment {
-		// Case 2b or 2c: The zone does not exist, or was not created by an older version of this CLI.
-		// Get the NS records for the delegation set (using the existing zone) and let Pulumi create the hosted zone for us
-		var zoneId *string
-		if zone != nil {
-			zoneId = zone.Id
+		// Check if the zone was created by the older CLI (before the delegation set was introduced)
+		if zone.Config.Comment != nil && *zone.Config.Comment == aws.CreateHostedZoneCommentLegacy {
+			// Case 2a: The zone was created by the older CLI, we'll use the existing NS records; track how many times this happens
+			track.Evt("Compose-Up delegateSubdomain old", track.P("domain", projectDomain))
+			return nsServers, "", nil
 		}
-		// TODO: avoid creating the delegation set if we're in preview mode
-		delegationSet, err := aws.CreateDelegationSet(ctx, zoneId, r53Client)
-		var delegationSetAlreadyCreated *r53types.DelegationSetAlreadyCreated
-		var delegationSetAlreadyReusable *r53types.DelegationSetAlreadyReusable
-		if errors.As(err, &delegationSetAlreadyCreated) || errors.As(err, &delegationSetAlreadyReusable) {
+
+		// Case 2b or 2c: The zone was not created by an older version of this CLI. We'll get the delegation set and let CD/Pulumi create/update the hosted zone
+		// TODO: we need to detect the case 2c where the zone was created by another tool and we need to use the existing NS records
```

Comment:
> TODO: we need to detect the case 2c where the zone was created by another tool

For my understanding: At this point in the code, we're sure the zone exists, but we don't know if we created it or if another tool created it. Is that correct?

Reply:
yes. In case Pulumi created it, then the Pulumi code will be "recreating" the zone instance, and update the AWS zone accordingly. In case the zone was created by a 3rd party tool or manually, Pulumi update will likely fail. It's a possibility, but not important enough to fix until someone actually runs into it.